Add GraphQL Yoga on Cloudflare Workers blog post (#1263)
the-guild-org/website: website/pages/blog/graphql-yoga-worker.mdx

support for connections over TCP.

https://blog.cloudflare.com/workers-tcp-socket-api-connect-databases/

I think that just because it is possible to connect via TCP, the advantage of the edge location is
lost a bit, because if an application running on the edge connects to a database in another location
to retrieve data, there is an overhead for that connection. So I'm waiting with anticipation for the
[Cloudflare D1](https://developers.cloudflare.com/d1/) to come out in the future, as it will
maximize the advantages of that location. Cloudflare Workers is as great a service as Cloud Run, so
the purpose of this article is to actually run it as a backend API server on top of Cloudflare
Workers.

## Architectures

Let's proceed under the assumption that the backend API will be prepared as a GraphQL server, and I
will write down what to use to build it.

### Cloudfalre Workers

https://workers.cloudflare.com/

The first assumption is this. The choice of Cloudflare Workers as a service or runtime narrows the
technologies that can be used in an application. Therefore, it is written as an assumption.

### Web Server

https://the-guild.dev/graphql/yoga-server

> Web Servers that work with Cloudflare Workers are pretty limited. This is because Node.js does not

Reviewer: ```suggestion
Many existing/older Web Servers have limited support for Cloudflare Workers. That's because they were built at a time to only support Node.js. 

Newer Web Server frameworks like [feTS](https://the-guild.dev/openapi/fets/server/quick-start) or [Hono](https://hono.dev/) are modern and have better support for Cloudflare Workers.

In our case, we will use [GraphQL Yoga](https://the-guild.dev/graphql/yoga-server) because it support GraphQL and also uses the same powerful [@whatwg-node library](https://github.com/ardatan/whatwg-node) as [feTS](https://the-guild.dev/openapi/fets/server/quick-start), to support [any Javascript runtime](https://the-guild.dev/blog/fetch-for-servers), including Cloudflare Workers.
```

> work. The typical one is [Hono](https://hono.dev/), but in this case we will choose

Reviewer 2: ```suggestion
work. Usually the preferred HTTP frameworks are [feTS](https://the-guild.dev/openapi/fets/server/quick-start) or [Hono](https://hono.dev/), but in this case we will choose
```

Can we also mention [feTS Server](https://the-guild.dev/openapi/fets/server/quick-start)?

+ export const connection = () => new Kysely<DB>({
```

My code is defined in a function and each call is an instantiation of Kysely. This is the intention.
The reason is that when connecting over TCP from Cloudflare Workers, TCP disconnects on a
per-request basis, so I need to make a connection to the database on a per-request basis. Kysely
uses something called `Pool` in the `pg` it uses for connections, which does connection pooling.
Kysely also only accepts `PostgresPool` connections, so this is what it does. connection pooling is
simply explained as defining multiple connections from one application to a database, and using
those connections within an application. However, this connection pooling is incompatible with the
Cloudflare Workers specification that TCP connections are broken after each request. In other words,
when a second request is received, the connection object is already connected, but because
Cloudflare Workers' TCP connection is broken, it cannot connect to the database and an error occurs.
So the code dares to instantiate Kysely in order to reconnect. The following configuration is
required for `pg` to work on Cloudflare Workers.

> ```toml
> # wrangler.toml

Reviewer 3: ````suggestion
```toml filename="wrangler.toml"
````

node_compat = true
```

If you do not enable this option, pg will not work.

https://developers.cloudflare.com/workers/wrangler/configuration/#node-compatibility

### Create GraphQL schema

Finally, write the GraphQL code to fetch data from the database and return it to the client. Some of
the code was [suggested](https://github.com/chimame/graphql-yoga-worker-with-pothos/pull/1) by
[hayes](https://github.com/hayes), the author of `Pothos`, while I was working on this article.

```ts
// src/models/index.ts
export * from './User'
```

> ```ts
> // src/models/User.ts

Reviewer 3: ````suggestion
```ts filename="src/models/User.ts"
````

import { AllSelection } from 'kysely/dist/cjs/parser/select-parser'
import { builder } from '../builder'
import { DB } from '../types'

export const UserType = builder.objectRef<AllSelection<DB, 'User'>>('User')

UserType.implement({
  fields: t => ({
    id: t.exposeID('id'),
    name: t.exposeString('name', { nullable: true }),
    email: t.exposeString('email'),
    initial: t.string({
      nullable: true,
      resolve: user => user.name?.slice(0, 1)
    })
  })
})
```

```ts
// src/resolvers/index.ts
export * from './User'
```

> ```ts
> // src/resolvers/Users/index.ts

Reviewer 3: ````suggestion
```ts filename="src/resolvers/Users/index.ts"
````

export * from './query'
```

> ```ts
> // src/resolvers/User/query.ts

Reviewer 3: ````suggestion
```ts filename="src/resolvers/User/query.ts"
````

import { builder } from '../../builder'
import { UserType } from '../../models/User'
import { connection } from '../../context'

builder.queryFields(t => ({
  User: t.field({
    type: UserType,
    nullable: true,
    resolve: async () => {
      const db = connection()
      await db.selectFrom('User').selectAll().executeTakeFirst()
    }
  })
}))
```

> ```diff
> // src/schema.ts

Reviewer 3: ````suggestion
```diff filename="src/schema.ts"
````
